### PR TITLE
Add Auto Encoder Model and Evaluation Notebook

### PR DESCRIPTION
**Pull request by asyatrhl**

### CODE REVIEW

File: datasets/samplemotordatalimerick.py [hunk 1]
```diff
@@ -0,0 +1,662 @@
...
+
+import pandas as pd
+import scipy
+
+import ai8x
+
+
+class SampleMotorDataLimerick(Dataset):
+    """
+    ADXL356C
```
Comment (seldauyanik-maxim) [inline datasets/samplemotordatalimerick.py:29]: We can update the docstring, can also be done after dataset repo is opened

File: datasets/samplemotordatalimerick.py [hunk 2]
```diff
@@ -0,0 +1,662 @@
...
+
+        self.__makedir_exist_ok(processed_folder)
+
+        self.specs_identifier = f'eval_mode_{self.eval_mode}_' + \
+                                f'label_as_signal_{self.label_as_signal}_' + \
+                                f'ds_{self.downsampling_ratio}_' + \
+                                f'dur_{self.signal_duration_in_sec}_' + \
+                                f'ovlp_ratio_{self.overlap_ratio}_' + \
+                                f'random_split_{self.random_or_speed_split}_' + \
+                                f'sensor_selected_{self.sensor_selected}_' +\
```
Comment (seldauyanik-maxim) [inline datasets/samplemotordatalimerick.py:261]: no need to log sensor option as identifier

File: datasets/samplemotordatalimerick.py [hunk 3]
```diff
@@ -0,0 +1,662 @@
...
+    signal_duration_in_sec = 0.25
+    overlap_ratio = 0.75
+
+    wanted_sampling_rate_Hz = 2000
+    downsampling_ratio = round(SampleMotorDataLimerick.sensor_options_sr_Hz[selected_sensor_idx] /
+                               wanted_sampling_rate_Hz)
+
+    # ds_ratio = 10,  sr: 20K / 10 = 2000, 0.25 sec window, fft input will have: 500 samples,
+    # fftout's first 256 samples will be used
+    # cnn input will have 2556 samples
```
Comment (seldauyanik-maxim) [inline datasets/samplemotordatalimerick.py:626]: typo: 2556 -> 256

File: scripts/evaluate_autoencoder.sh [hunk 4]
```diff
@@ -0,0 +1,2 @@
+#!/bin/sh
+python train.py --deterministic --model ai85autoencoder --dataset SampleMotorDataLimerick_ForEvalWithSignal --regression --device MAX78000 --qat-policy policies/qat_policy_autoencoder.yaml --use-bias --evaluate --exp-load-weights-from ../ai8x-synthesis/trained/ae_cork_localminmax_random_qat-q.pth.tar -8 --print-freq 1 "$@"
```
Comment (seldauyanik-maxim) [inline scripts/evaluate_autoencoder.sh:2]: Better to rename trained model: replace cork with limerick, no need for random and localminmax

File: datasets/samplemotordatalimerick.py [hunk 5]
```diff
@@ -0,0 +1,662 @@
...
+        return features
+
+    def __gen_datasets(self):
+        print('\nGenerating dataset pickle files from the raw data '
+              f'files (specs identifier: {self.specs_identifier}) ...\n')
+
+        actual_root_dir = os.path.join(self.root, self.__class__.__name__,
+                                       "SpectraQuest_Rig_Data_Voyager_3/")
+
+        data_dir = os.path.join(actual_root_dir, f'Test_Results_Data_{self.sensor_selected}/')
```
Comment (seldauyanik-maxim) [inline datasets/samplemotordatalimerick.py:318]: Most probably the csv files folder will not have Test_Results_ prefix

Comment (seldauyanik-maxim) [reply datasets/samplemotordatalimerick.py:318]: Also a check here wrt data_dir would be nice to add, can also be added after dataset repo is published. If folder does not exist or empty data loader will return 0 samples wo any control

File: datasets/samplemotordatalimerick.py [hunk 6]
```diff
@@ -0,0 +1,662 @@
...
+    """
+    ADXL356C
+    """
+
+    # Order 0 is used for default sensor: ADX365C
+    sensor_options = ['ADXL356C']
+    # Order 0 is reserved for 'all' do not change order
+    rpm_options = ('all', '0600', '1200', '1800', '2400', '3000')
+    # Order 0 is used for default sensor: ADX365C, do not change order
+    sensor_options_sr_Hz = [20000]
```
Comment (seldauyanik-maxim) [inline datasets/samplemotordatalimerick.py:37]: we can remove all options and keep sampling rate and data len as variables

File: datasets/samplemotordatalimerick.py (hunk 1 above)
Comment (asyatrhl) [reply datasets/samplemotordatalimerick.py]: Updated

File: datasets/samplemotordatalimerick.py (hunk 2 above)
Comment (asyatrhl) [reply datasets/samplemotordatalimerick.py]: Updated

File: scripts/evaluate_autoencoder.sh (hunk 4 above)
Comment (asyatrhl) [reply scripts/evaluate_autoencoder.sh]: Updated

File: datasets/samplemotordatalimerick.py (hunk 5 above)
Comment (asyatrhl) [reply datasets/samplemotordatalimerick.py]: Updated

File: datasets/samplemotordatalimerick.py [hunk 7]
```diff
@@ -0,0 +1,692 @@
+###################################################################################################
+#
+# Copyright (C) 2024 Analog Devices, Inc. All Rights Reserved.
+# This software is proprietary to Analog Devices, Inc. and its licensors.
+#
+###################################################################################################
+"""
+Classes and functions for the Limerick Sample Motor Data Dataset
```
Comment (rotx-maxim) [inline datasets/samplemotordatalimerick.py:8]: Please add a link to the documentation here

File: notebooks/AutoEncoder_Evaluation.ipynb [hunk 8]
```diff
@@ -0,0 +1,565 @@
...
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "###################################################################################################\n",
+    "#\n",
+    "# Copyright (C) 2023 Analog Devices, Inc. All Rights Reserved.\n",
+    "# This software is proprietary and confidential to Analog Devices, Inc. and its licensors.\n",
+    "#\n",
+    "###################################################################################################\n",
+    "import os\n",
```
Comment (rotx-maxim) [inline notebooks/AutoEncoder_Evaluation.ipynb:15]: Please add a link to the documentation at the top of the notebook

File: notebooks/AutoEncoder_Evaluation.ipynb [hunk 9]
```diff
@@ -0,0 +1,565 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Change this checkpoint file path with your own trained one\n",
+    "checkpoint_path = '../ai8x-synthesis/trained/ai85-autoencoder-samplemotordatalimerick-qat.pth.tar'"
```
Comment (rotx-maxim) [inline notebooks/AutoEncoder_Evaluation.ipynb:165]: please use os.path.join

File: notebooks/AutoEncoder_Evaluation.ipynb [hunk 10]
```diff
@@ -0,0 +1,565 @@
...
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "{'start_epoch': 200, 'weight_bits': 8, 'shift_quantile': 0.995}\n",
+      "Configuring device: MAX78000, simulate=False.\n"
+     ]
+    }
+   ],
+   "source": [
+    "qat_yaml_file_used_in_training = '../policies/qat_policy_autoencoder.yaml'\n",
```
Comment (rotx-maxim) [inline notebooks/AutoEncoder_Evaluation.ipynb:183]: os.path.join

File: notebooks/AutoEncoder_Evaluation.ipynb [hunk 11]
```diff
@@ -0,0 +1,565 @@
...
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Configuring device: MAX78000, simulate=False.\n"
+     ]
+    }
+   ],
+   "source": [
+    "device = torch.device(\"cuda:1\" if torch.cuda.is_available() else \"cpu\")\n",
+    "data_path = '../data'\n",
```
Comment (rotx-maxim) [inline notebooks/AutoEncoder_Evaluation.ipynb:73]: os.path.join

**Review by rotx-maxim — CHANGES_REQUESTED** [review]
Please add links to the documentation and dataset repository, and use os.path.join instead of "/" or "../" (affects both code and notebook)

File: datasets/samplemotordatalimerick.py (hunk 6 above)
Comment (asyatrhl) [reply datasets/samplemotordatalimerick.py]: Sensor and rpm selection is removed

File: datasets/samplemotordatalimerick.py (hunk 7 above)
Comment (asyatrhl) [reply datasets/samplemotordatalimerick.py:8]: Documentation is currently on pr, link will be added when it is published.

File: notebooks/AutoEncoder_Evaluation.ipynb (hunk 8 above)
Comment (asyatrhl) [reply notebooks/AutoEncoder_Evaluation.ipynb:15]: Documentation is currently on pr, link will be added when it is published.

File: notebooks/AutoEncoder_Evaluation.ipynb (hunk 9 above)
Comment (asyatrhl) [reply notebooks/AutoEncoder_Evaluation.ipynb]: Updated

File: notebooks/AutoEncoder_Evaluation.ipynb (hunk 10 above)
Comment (asyatrhl) [reply notebooks/AutoEncoder_Evaluation.ipynb]: Updated

File: notebooks/AutoEncoder_Evaluation.ipynb (hunk 11 above)
Comment (asyatrhl) [reply notebooks/AutoEncoder_Evaluation.ipynb]: Updated

File: notebooks/AutoEncoder_Evaluation.ipynb (hunk 8 above)
Comment (rotx-maxim) [reply notebooks/AutoEncoder_Evaluation.ipynb:15]: We already know the location of the documentation, please add it now so we don't need any further changes

Comment (asyatrhl) [reply notebooks/AutoEncoder_Evaluation.ipynb:20]: Updated

**Review by rotx-maxim — APPROVED** [review]

File: datasets/samplemotordatalimerick.py [hunk 12]
```diff
@@ -0,0 +1,671 @@
...
+
+class SampleMotorDataLimerick(Dataset):
+    """
+    Sample motor data is collected using SpectraQuest Machinery Fault Simulator.
+    ADXL356 sensor data is used for vibration raw data.
+    For ADXL356 sensor, the sampling frequency was 20kHz and
+    data csv files recorded for 2 sec in X, Y and Z direction.
+    """
+
+    # Order 0 is reserved for 'all' do not change order
```
Comment (seldauyanik-maxim) [inline datasets/samplemotordatalimerick.py:38]: Better to remove above irrelevant comment

File: datasets/samplemotordatalimerick.py [hunk 13]
```diff
@@ -0,0 +1,671 @@
...
+
+def samplemotordatalimerick_get_datasets_for_eval_with_signal(data,
+                                                              load_train=True,
+                                                              load_test=True):
+    """"
+    Returns Sample Motor Data Limerick Dataset For Evaluation Mode
+    Label is signal
+    """
+
+    eval_mode = True   # Test set includes validation normals
```
Comment (seldauyanik-maxim) [inline datasets/samplemotordatalimerick.py:623]: comment would better be: testset now includes anormal samples as well as validation normals

**Review by seldauyanik-maxim — APPROVED** [review]

**Review by rotx-maxim — APPROVED** [review]

**Pull request merged**